packages/react-jss/src/utils/sheets.js

```diff
@@ -96,8 +96,10 @@ export const updateDynamicRules = (data: any, sheet: StyleSheet, rules: DynamicR
   // Loop over each dynamic rule and update it
   // We can't just update the whole sheet as this has all of the rules for every component instance
   for (const key in rules) {
```

Comment:
shouldn't we rather not have that key in rules?

Reply:
Ideally yes. The current code relying on `isFirstRender` is not really fixable as hot-reloading unmounts and remounts the component but preserves state, meaning effect cleanups are ran as if the component was unmounted, but `isFirstRender` is still true.

I tried to write a more idiomatic and cleaner version that works independently of the first render. The commit is here: https://github.com/bs85/jss/commit/7fbb7b9775baa5cf1631f91cac9d47944acaf007

Unfortunately it breaks ~10 tests, and I don't have the resources to dig deeper.

If you want to spend time on this feel free to do it, otherwise my quick fix would be nice for now.

Reply:
I would really love to have a proper fix instead of a workaround

Reply:
So do I...

Here's the logs from running the tests, as you're more familiar, any idea what it could be?

```
FAILED TESTS:
    ✖ should render a number of composed style rules with styled API
      Chrome 78.0.3904 (Linux 0.0.0)
    Error: expected '.scd-0 {}\n.scd-0-1 {\n  color: white;\n  font-size: 32px;\n  font-weight: bold;\n  padding-top: 4px;\n  padding-left: 16px;\n  padding-right: 16px;\n  padding-bottom: 4px;\n  background-color: #07c;\n}\n@media screen and (min-width: 40em) {\n  .scd-0-1 {\n    font-size: 48px;\n    padding-top: 8px;\n    padding-left: 32px;\n    padding-right: 32px;\n    padding-bottom: 8px;\n  }\n}\n@media screen and (min-width: 52em) {\n  .scd-0-1 {\n    font-size: 64px;\n  }\n}\n@media screen and (min-width: 40em) {\n  .scd-0-1 {\n    font-size: 48px;\n    padding-top: 8px;\n    padding-left: 32px;\n    padding-right: 32px;\n    padding-bottom: 8px;\n  }\n}\n@media screen and (min-width: 52em) {\n  .scd-0-1 {\n    font-size: 64px;\n  }\n}' to equal '.scd-0 {}\n.scd-0-1 {\n  color: white;\n  font-size: 32px;\n  font-weight: bold;\n  padding-top: 4px;\n  padding-left: 16px;\n  padding-right: 16px;\n  padding-bottom: 4px;\n  background-color: #07c;\n}\n@media screen and (min-width: 40em) {\n  .scd-0-1 {\n    font-size: 48px;\n    padding-top: 8px;\n    padding-left: 32px;\n    padding-right: 32px;\n    padding-bottom: 8px;\n  }\n}\n@media screen and (min-width: 52em) {\n  .scd-0-1 {\n    font-size: 64px;\n  }\n}'
        at Assertion.assert (webpack://jss/node_modules/expect.js/index.js:96:0 <- packages/react-jss/tests/styledSystem.js:11097:13)
        at Assertion.be.Assertion.equal (webpack://jss/node_modules/expect.js/index.js:216:0 <- packages/react-jss/tests/styledSystem.js:11217:10)
        at Assertion.<computed> [as be] (webpack://jss/node_modules/expect.js/index.js:69:0 <- packages/react-jss/tests/styledSystem.js:11070:24)
        at Context.<anonymous> (webpack://jss/packages/react-jss/tests/styledSystem.js:87:35 <- packages/react-jss/tests/styledSystem.js:30300:78)

  React-JSS: theming useStyles()
    ✖ one themed instance wo/ = 1 style, theme update = 1 style
      Chrome 78.0.3904 (Linux 0.0.0)
    Error: expected true to equal false
        at Assertion.assert (webpack://jss/node_modules/expect.js/index.js:96:0 <- packages/react-jss/tests/useStylesTheming.js:11097:13)
        at Assertion.be.Assertion.equal (webpack://jss/node_modules/expect.js/index.js:216:0 <- packages/react-jss/tests/useStylesTheming.js:11217:10)
        at Assertion.<computed> [as be] (webpack://jss/node_modules/expect.js/index.js:69:0 <- packages/react-jss/tests/useStylesTheming.js:11070:24)
        at Context.<anonymous> (webpack://jss/packages/react-jss/tests/useStylesTheming.js:165:45 <- packages/react-jss/tests/useStylesTheming.js:29233:88)

    ✖ one themed instance w/ dynamic props = 2 styles, theme update = 2 styles
      Chrome 78.0.3904 (Linux 0.0.0)
    Error: expected true to equal false
        at Assertion.assert (webpack://jss/node_modules/expect.js/index.js:96:0 <- packages/react-jss/tests/useStylesTheming.js:11097:13)
        at Assertion.be.Assertion.equal (webpack://jss/node_modules/expect.js/index.js:216:0 <- packages/react-jss/tests/useStylesTheming.js:11217:10)
        at Assertion.<computed> [as be] (webpack://jss/node_modules/expect.js/index.js:69:0 <- packages/react-jss/tests/useStylesTheming.js:11070:24)
        at Context.<anonymous> (webpack://jss/packages/react-jss/tests/useStylesTheming.js:189:45 <- packages/react-jss/tests/useStylesTheming.js:29253:88)

    ✖ two themed instances w/ dynamic props w/ same theme = 3 styles, theme update = 3 styles
      Chrome 78.0.3904 (Linux 0.0.0)
    Error: expected true to equal false
        at Assertion.assert (webpack://jss/node_modules/expect.js/index.js:96:0 <- packages/react-jss/tests/useStylesTheming.js:11097:13)
        at Assertion.be.Assertion.equal (webpack://jss/node_modules/expect.js/index.js:216:0 <- packages/react-jss/tests/useStylesTheming.js:11217:10)
        at Context.<anonymous> (webpack://jss/packages/react-jss/tests/useStylesTheming.js:243:45 <- packages/react-jss/tests/useStylesTheming.js:29299:88)

    ✖ two themed instances wo/ dynamic props w/ same theme = 1 styles, different theme update = 2 styles
      Chrome 78.0.3904 (Linux 0.0.0)
    Error: expected 1 to equal 2
        at Assertion.assert (webpack://jss/node_modules/expect.js/index.js:96:0 <- packages/react-jss/tests/useStylesTheming.js:11097:13)
        at Assertion.be.Assertion.equal (webpack://jss/node_modules/expect.js/index.js:216:0 <- packages/react-jss/tests/useStylesTheming.js:11217:10)
        at Context.<anonymous> (webpack://jss/packages/react-jss/tests/useStylesTheming.js:273:40 <- packages/react-jss/tests/useStylesTheming.js:29319:83)

    ✖ two themed instances w/ dynamic props w/ same theme = 3 styles, different theme update = 4 styles
      Chrome 78.0.3904 (Linux 0.0.0)
    Error: expected 1 to equal 2
        at Assertion.assert (webpack://jss/node_modules/expect.js/index.js:96:0 <- packages/react-jss/tests/useStylesTheming.js:11097:13)
        at Assertion.be.Assertion.equal (webpack://jss/node_modules/expect.js/index.js:216:0 <- packages/react-jss/tests/useStylesTheming.js:11217:10)
        at Context.<anonymous> (webpack://jss/packages/react-jss/tests/useStylesTheming.js:302:40 <- packages/react-jss/tests/useStylesTheming.js:29346:83)

    ✖ two themed instances wo/ dynamic props w/ different themes = 2 styles, same theme update = 1 style
      Chrome 78.0.3904 (Linux 0.0.0)
    Error: expected true to equal false
        at Assertion.assert (webpack://jss/node_modules/expect.js/index.js:96:0 <- packages/react-jss/tests/useStylesTheming.js:11097:13)
        at Assertion.be.Assertion.equal (webpack://jss/node_modules/expect.js/index.js:216:0 <- packages/react-jss/tests/useStylesTheming.js:11217:10)
        at Context.<anonymous> (webpack://jss/packages/react-jss/tests/useStylesTheming.js:331:45 <- packages/react-jss/tests/useStylesTheming.js:29365:88)

    ✖ two themed instances w/ dynamic props w/ different themes = 4 styles, same theme update = 3 styles
      Chrome 78.0.3904 (Linux 0.0.0)
    Error: expected true to equal false
        at Assertion.assert (webpack://jss/node_modules/expect.js/index.js:96:0 <- packages/react-jss/tests/useStylesTheming.js:11097:13)
        at Assertion.be.Assertion.equal (webpack://jss/node_modules/expect.js/index.js:216:0 <- packages/react-jss/tests/useStylesTheming.js:11217:10)
        at Context.<anonymous> (webpack://jss/packages/react-jss/tests/useStylesTheming.js:361:45 <- packages/react-jss/tests/useStylesTheming.js:29393:88)
```